sdk/core/azure-core/src/logging/logging.cpp

```diff
@@ -7,72 +7,41 @@
 #include <mutex>
 
 using namespace Azure::Core::Logging;
-using namespace Azure::Core::Logging::Details;
-
-class Azure::Core::Logging::Details::LogClassificationsPrivate {
-public:
-  static LogClassifications const LogClassificationsConstant(bool all)
-  {
-    return LogClassifications(all);
-  }
-
-  static bool IsClassificationEnabled(LogClassifications const& cls, LogClassification const& c)
-  {
-    return cls.m_all || (cls.m_classifications.find(c) != cls.m_classifications.end());
-  }
-};
+using namespace Azure::Core::Logging::Internal;
 
 namespace {
 std::mutex g_loggerMutex;
 LogListener g_logListener(nullptr);
+LogLevel g_logLevel = LogLevel::Verbose;
```

Comment:
The default is either Error or maybe we need a None.
Also, this variable should be atomic.

Reply:
We need to sync both level and callback, std::function is not atomic. Other languages don't have None. If you want none, why have callback set?

Reply:
If both variables are atomic, then you don't need the mutex.
The customer needs to provide a callback so they can write to the console, file, etc.
The callback can be set and then the level can be adjusted dynamically over the lifetime of the app.

Reply:
`std::atomic<std::function<...>>` does not compile, underlying type does not correspond to atomic requirements.

Reply:
Could make it an atomic void* and then cast.

Reply:
I thought about that, I am not sure if std::function's pointer can always be represented as void*, or maybe under some circumstances it could have more. Because I'd expect `std::atomic<>` to also compile if the type is compatible, but apparently it's not. I'll talk to Casey.If we want this checked in today, let's defer this decision because I am not sure it is always safe to cast. It won't be a breaking change to update it in the future, it is an implementation detail, only visible there.

Reply:
@CaseyCarter, what are our options?

Reply:
I agree Anton - can be changed later.

Reply:
`std::function` can represent function objects with state, for which there is no corresponding equivalent pointer-to-function. `std::atomic` requires its parameter to be trivially copyable - effectively a bag of bits that can be `memcpy`d - which `std::function` is not (if there's more state than will fit in the small object buffer inside the `std::function` it has to allocate memory).

Without having scrutinized the code (or looked at anything outside of this comment) why do we hate `mutex`?

Reply:
I wouldn't say we "hate mutex". Keeping a mutex is fine. It's just that atomic is usually faster - but this is unlikely to be a performance-sensitive piece of code..